source/common/config/incremental_subscription_impl.h

```diff
@@ -0,0 +1,214 @@
+#pragma once
+
+#include <queue>
+
+#include "envoy/api/v2/discovery.pb.h"
+#include "envoy/common/token_bucket.h"
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/backoff_strategy.h"
+#include "common/common/logger.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/grpc_stream.h"
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+struct ResourceNameDiff {
+  std::vector<std::string> added_;
+  std::vector<std::string> removed_;
+};
+
+/**
+ * TODO SOMETHING SOMETHING subscription. Also handles per-xDS API stats/logging.
+ */
+template <class ResourceType>
+class IncrementalSubscriptionImpl
+    : public IncrementalSubscription<ResourceType>,
+      public GrpcStream<envoy::api::v2::IncrementalDiscoveryRequest,
+                        envoy::api::v2::IncrementalDiscoveryResponse, ResourceNameDiff> {
+public:
+  IncrementalSubscriptionImpl(const LocalInfo::LocalInfo& local_info,
+                              Grpc::AsyncClientPtr async_client, Event::Dispatcher& dispatcher,
+                              const Protobuf::MethodDescriptor& service_method,
+                              Runtime::RandomGenerator& random, Stats::Scope& scope,
+                              const RateLimitSettings& rate_limit_settings,
+                              IncrementalSubscriptionStats stats)
+      : GrpcStream<envoy::api::v2::IncrementalDiscoveryRequest,
+                   envoy::api::v2::IncrementalDiscoveryResponse, ResourceNameDiff>(
+            std::move(async_client), service_method, random, dispatcher, scope,
+            rate_limit_settings),
+        type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())),
+        stats_(stats) {
+    request_.set_type_url(type_url_);
+    request_.mutable_node()->MergeFrom(local_info.node());
+    establishNewStream();
```

Comment:
establishNewStream() here starts a new stream, but then another exchange follows when ```start(...)``` (https://github.com/envoyproxy/envoy/pull/5466/files#diff-691e22a9e24880645b6ad529b3a3470aR181) is called in the ```initialize``` callback here: https://github.com/envoyproxy/envoy/pull/5466/files#diff-1b7e10977ff1c16c485aa6472d859ccbR29. I'm not sure if this could be a problem in cds (it is in rds: I may not have fields initialized that must be present in a non-initial request/response cycle). Could we skip ```establishNewStream``` call here, and call it in ```start``` instead? This would reduce connection setup to one request/response exchange, and as a bonus ```start``` method would lose```resources``` argument...

Reply:
Sure, that makes sense.

Reply:
Unfortunately this change makes the integration test fail. I think it's just a matter of the order that things happen within the implementation, and with some work they could be reordered without breaking anything.

Reply:
(fixed after discussion on slack)